tonic/src/codec/compression.rs

```diff
@@ -10,62 +10,79 @@ pub(crate) const ENCODING_HEADER: &str = "grpc-encoding";
 pub(crate) const ACCEPT_ENCODING_HEADER: &str = "grpc-accept-encoding";
 
 /// Struct used to configure which encodings are enabled on a server or channel.
-#[derive(Debug, Default, Clone, Copy)]
+///
+/// Represents an ordered list of compression encodings that are enabled.
+#[derive(Debug, Clone, Copy)]
 pub struct EnabledCompressionEncodings {
-    #[cfg(feature = "gzip")]
-    pub(crate) gzip: bool,
-    #[cfg(feature = "zstd")]
-    pub(crate) zstd: bool,
+    inner: [Option<CompressionEncoding>; 2],
 }
 
 impl EnabledCompressionEncodings {
-    /// Check if a [`CompressionEncoding`] is enabled.
-    pub fn is_enabled(&self, encoding: CompressionEncoding) -> bool {
-        match encoding {
-            #[cfg(feature = "gzip")]
-            CompressionEncoding::Gzip => self.gzip,
-            #[cfg(feature = "zstd")]
-            CompressionEncoding::Zstd => self.zstd,
+    /// Enable a [`CompressionEncoding`].
+    ///
+    /// Adds the new encoding to the end of the encoding list.
+    pub fn enable(&mut self, encoding: CompressionEncoding) {
+        for e in self.inner.iter_mut() {
+            match e {
+                Some(e) if *e == encoding => return,
+                None => {
+                    *e = Some(encoding);
+                    return;
+                }
+                _ => continue,
+            }
         }
     }
 
-    /// Enable a [`CompressionEncoding`].
-    pub fn enable(&mut self, encoding: CompressionEncoding) {
-        match encoding {
-            #[cfg(feature = "gzip")]
-            CompressionEncoding::Gzip => self.gzip = true,
-            #[cfg(feature = "zstd")]
-            CompressionEncoding::Zstd => self.zstd = true,
+    /// Remove the last [`CompressionEncoding`].
+    pub fn pop(&mut self) -> Option<CompressionEncoding> {
+        for entry in self.inner.iter_mut().rev() {
+            match entry {
+                Some(e) => {
+                    let removed = *e;
+                    *entry = None;
+                    return Some(removed);
+                }
+                _ => continue,
+            }
         }
+
+        None
     }
 
     pub(crate) fn into_accept_encoding_header_value(self) -> Option<http::HeaderValue> {
-        match (self.is_gzip_enabled(), self.is_zstd_enabled()) {
-            (true, false) => Some(http::HeaderValue::from_static("gzip,identity")),
-            (false, true) => Some(http::HeaderValue::from_static("zstd,identity")),
-            (true, true) => Some(http::HeaderValue::from_static("gzip,zstd,identity")),
-            (false, false) => None,
+        let mut value = BytesMut::new();
+        for encoding in self.inner.iter().copied().filter_map(|e| e) {
```

Comment:
The value copying seems not to be needed. `filter_map` is used just for filtering without mapping of the values, I would think `flatten` might be idiomatic in this case.

```suggestion
        for encoding in self.inner.into_iter().flatten() {
```

Reply:
I don't like `flatten()` in this kind of context, but took your suggestion to use `into_iter()`.

Reply:
Why would not you like `flatten()` here?

My opinion is that it would be better to follow the recommendations of the Rust documentation and tooling.

The official document describes it as the typical use case.

https://doc.rust-lang.org/std/iter/trait.Iterator.html#method.flatten

And clippy warns it.

https://rust-lang.github.io/rust-clippy/master/index.html#/filter_map_identity

Reply:
Okay, changed it. I don't like `flatten()` because IMO it's counterintuitive to treat `Option` as an iterator.